File: nodejs/packages/node-red-contrib-periph-power/nodes/ina219/ina219.js
'use strict';

module.exports = function(RED) {
    const { I2CConnection } = require('periph/src/connection/i2c');
    const { INA219Full }   = require('periph/src/chips/power/ina219');

    function INA219DeviceNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        try {
            const connection = new I2CConnection(parseInt(config.bus), parseInt(config.address, 16));
            node.driver     = new INA219Full(connection, parseFloat(config.rShunt), parseFloat(config.maxCurrent));
            node.connection = connection;
        } catch (e) {
            node.error('INA219 init failed: ' + e.message);
        }
        node.on('close', function() {
            if (node.connection) node.connection.close();
        });
    }
    RED.nodes.registerType('ina219-device', INA219DeviceNode);

    function INA219ReadNode(config) {
        RED.nodes.createNode(this, config);
        const node = this;
        node.device = RED.nodes.getNode(config.device);

        node.on('input', async function(msg, send, done) {
            if (!node.device || !node.device.driver) {
                node.error('No INA219 device configured', msg);
                done();
                return;
            }
            try {
                const d = node.device.driver;
                msg.payload = {
                    voltage:      await d.voltage(),
                    current:      await d.current(),
                    power:        await d.power(),
                    shuntVoltage: await d.shuntVoltage()
                };
                send(msg);
                done();
            } catch (e) {
                done(e);
            }
        });
    }
    RED.nodes.registerType('ina219', INA219ReadNode);
};

File: python/uiflow2/power/ina219/INA219.py
"""
file     INA219
time     2026-08-27
author
email
license  Apache License 2.0
"""

from periph.connection.i2c_auto import I2CConnection
from periph.chips.power.ina219 import INA219Full


class INA219:
    """
    note:
        en: ''
    details:
        color: '#C084FC'
        link: https://github.com/tuhde/Periph
        image: ''
        category: Custom
    example: ''
    """

    def __init__(self, bus: int = 0, address: int = 64, r_shunt: float = 0.1, max_current: float = 2.0):
        """
        label:
            en: '%1 init bus %2 address %3 r_shunt %4 max_current %5'
        params:
            bus:
                name: bus
                type: int
                default: '0'
                field: number
            address:
                name: address
                type: int
                default: '64'
                field: number
            r_shunt:
                name: r_shunt
                type: float
                default: '0.1'
                field: number
            max_current:
                name: max_current
                type: float
                default: '2'
                field: number
        """
        connection = I2CConnection(address, bus=bus)
        self._driver = INA219Full(connection, r_shunt, max_current)

    def voltage(self) -> float:
        """
        label:
            en: '%1 bus voltage (V)'
        """
        return self._driver.voltage()

    def shunt_voltage(self) -> float:
        """
        label:
            en: '%1 shunt voltage (V)'
        """
        return self._driver.shunt_voltage()

    def current(self) -> float:
        """
        label:
            en: '%1 current (A)'
        """
        return self._driver.current()

    def power(self) -> float:
        """
        label:
            en: '%1 power (W)'
        """
        return self._driver.power()
